src/gluonts/mx/distribution/lds.py

```diff
@@ -211,33 +213,105 @@ def log_prob(
         Tensor
             Final mean, shape (batch_size, latent_dim)
         Tensor
             Final covariance, shape (batch_size, latent_dim, latent_dim)
         """
-        if scale is not None:
-            x = self.F.broadcast_div(x, scale.expand_dims(axis=1))
-        # TODO: Based on form of the prior decide to do either filtering
-        #   or residual-sum-of-squares
-        log_p, final_mean, final_cov = self.kalman_filter(x, observed)
-        if scale is not None:
-            F = self.F
-            # log_abs_det_jac: sum over all output dimensions.
-            ladj = -F.sum(F.log(F.abs(scale)), axis=-1, keepdims=True)
-
-            # Sum `ladj` over all time steps.
-            log_p = F.broadcast_add(log_p, ladj)
+        F = self.F
+        x_transformed, ladj = self.transform(x, scale, output_transform)
+        log_p, final_mean, final_cov = self.kalman_filter(
+            x_transformed, observed
+        )
+
+        if scale is not None or output_transform is not None:
+            # Get the shape right if only scaling is done.
+            ladj = F.broadcast_add(ladj, F.zeros_like(log_p))
+            # Mask `ladj` for missing values as these should not modify log prob.
+            ladj = F.where(
+                condition=observed,
+                x=ladj * observed,
+                y=self.F.zeros_like(ladj),
+            )
+            log_p = log_p + ladj
+
         return log_p, final_mean, final_cov
 
+    def transform(
+        self,
+        x: Tensor,
+        scale: Optional[Tensor] = None,
+        output_transform: Optional[ComposedBijectionHybridBlock] = None,
+    ):
+        """
+        Scales and then transforms the target `x` by the (inverse of) `output_transform`.
+
+        Parameters
+        ----------
+        x
+            Observations, shape (batch_size, seq_length, output_dim)
+        scale
+            Scale of each sequence in x, shape (batch_size, output_dim)
+        output_transform
+            Specifies the (inverse of) transformation to be applied to the target time series.
+            Given target `x` is first scaled then the inverse of `output_transform` is applied to make it more Gaussian-like.
+
+        Returns
+        -------
+        Tensor
+            Transformed target, shape (batch_size, seq_length, output_dim)
+            log_abs_det_jac,    shape (batch_size, 1) if `output_transform` is None, (batch_size, seq_length) otherwise
+        """
+        F = self.F
+
+        def _scale(x):
+            x_scaled, ladj = x, None
+            if scale is not None:
+                # We apply scaling before transforming the target by the given output transformation.
+                x_scaled = F.broadcast_div(x, scale.expand_dims(axis=1))
+
+                # log_abs_det_jac: sum over all output dimensions. Shape (batch_size, 1)
+                ladj = -F.sum(F.log(F.abs(scale)), axis=-1, keepdims=True)
```

Comment:
This could probably be done directly by using the AffineTransformation bijection which would handle the scale directly. One could also use a ComposedBijection object to compose the scale and the output_transform bijections.